src/HexagonAlignLoads.cpp

```diff
@@ -0,0 +1,242 @@
+                        AlignCheck ac = get_alignment_check(ramp, op->param.host_alignment(), nullptr);
+
+                        if (ac == AlignCheck::Unaligned) {
+                            debug(4) << "HexagonAlignLoads: base_a is unaligned: shifting base_b\n";
+                            debug(4) << "HexagonAlignLoads: Type: " << op->type << "\n";
+                            debug(4) << "HexagonAlignLoads: Index: " << index << "\n";
+                            base_b -= 1;
+                            b_shift = 1;
+                        }
+                    }
+
+                    Expr ramp_a = Ramp::make(base_a, 1, lanes);
+                    Expr ramp_b = Ramp::make(base_b, 1, lanes);
+                    Expr vec_a = mutate(Load::make(op->type, op->name, ramp_a, op->image, op->param));
+                    Expr vec_b = mutate(Load::make(op->type, op->name, ramp_b, op->image, op->param));
+
+                    std::vector<Expr> indices;
+                    for (int i = 0; i < lanes/2; ++i) {
+                        indices.push_back(i*2);
+                    }
+                    for (int i = lanes/2; i < lanes; ++i) {
+                        indices.push_back(i*2 + b_shift);
+                    }
+
+                    debug(4) << "HexagonAlignLoads: Unaligned Load: Converting " << (Expr) op << " into ...\n";
+
+                    expr = concat_and_shuffle(vec_a, vec_b, indices);
+
+                    debug(4) <<  "... " << expr << "\n";
+                    return;
+                }else {
+                    expr = op;
+                    return;
+                }
+            }
+        } else {
+            expr = op;
+            return;
+        }
+    }
+
+    template<typename NodeType, typename LetType>
+    void visit_let(NodeType &result, const LetType *op) {
+        Expr value;
+        if (op->value.type() == Int(32)) {
+            alignment_info.push(op->name, modulus_remainder(op->value, alignment_info));
+            value = op->value;
+        } else {
+            const Load *ld = op->value.template as<Load>();
+            if (ld) value = mutate(op->value);
+            else value = op->value;
+        }
+
+        NodeType body = mutate(op->body);
+        if (op->value.type() == Int(32)) {
+            alignment_info.pop(op->name);
+        }
+        result = LetType::make(op->name, value, body);
+    }
+
+    void visit(const Let *op) { visit_let(expr, op); }
+    void visit(const LetStmt *op) { visit_let(stmt, op); }
+    void visit(const For *op) {
+        if (op->device_api == DeviceAPI::Hexagon) {
+            if (target.has_feature(Target::HVX_128)) {
+                vector_size = 128;
+            }
+            else if (target.has_feature(Target::HVX_64)) {
+                vector_size = 64;
+            } else {
+                internal_error << "Unknown HVX mode";
+            }
+        }
+        int saved_vector_size = vector_size;
```

Comment:
Thinking about this more, in the offload case, we need to make this code only affect Hexagon code. Currently, it will run on the whole lowered func, even ARM code that is along side hexagon offloading. If it's at all possible, I think aligned loads and store forwarding should both be invoked from here: https://github.com/halide/Halide/blob/hexagon2/src/CodeGen_Hexagon.cpp#L104